assembly/src/lib.rs

```diff
@@ -1,45 +1,44 @@
-#![cfg_attr(not(feature = "std"), no_std)]
+#![no_std]
+#![cfg_attr(all(nightly, not(feature = "std")), feature(error_in_core))]
 
-#[cfg(not(feature = "std"))]
 #[macro_use]
 extern crate alloc;
+#[cfg(any(test, feature = "std"))]
+extern crate std;
 
 use vm_core::{
-    code_blocks::CodeBlock,
-    crypto,
+    crypto::hash::RpoDigest,
     errors::KernelError,
     utils::{
-        collections::{btree_map, BTreeMap},
         ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable, SliceReader,
     },
-    CodeBlockTable, Felt, Kernel, Operation, Program, StarkField, ONE, ZERO,
+    Felt, ONE, ZERO,
 };
 
-mod library;
-pub use library::{Library, LibraryNamespace, LibraryPath, MaslLibrary, Module, Version};
-
-mod procedures;
-use procedures::{CallSet, NamedProcedure, Procedure};
-pub use procedures::{ProcedureId, ProcedureName};
+#[cfg(feature = "formatter")]
+pub use vm_core::{prettier, utils::DisplayHex};
 
+mod assembler;
 pub mod ast;
-use ast::{NAMESPACE_LABEL_PARSER, PROCEDURE_LABEL_PARSER};
-
-mod tokens;
-use tokens::{Token, TokenStream};
-
+pub mod diagnostics;
 mod errors;
-pub use errors::{AssemblyError, LabelError, LibraryError, ParsingError, PathError};
-
-mod assembler;
-pub use assembler::{Assembler, AssemblyContext};
-
+pub mod library;
+mod parser;
+pub(crate) mod sema;
+#[cfg(any(test, feature = "testing"))]
+pub mod testing;
 #[cfg(test)]
 mod tests;
 
-// RE-EXPORTS
-// ================================================================================================
+pub use self::assembler::{ArtifactKind, Assembler, AssemblyContext};
+pub use self::ast::{Module, ProcedureName};
+pub use self::errors::AssemblyError;
+pub use self::library::{
+    Library, LibraryError, LibraryNamespace, LibraryPath, MaslLibrary, PathError, Version,
+};
+pub use self::parser::{SourceLocation, SourceSpan, Span, Spanned};
```

Comment:
> Why do we need to use self:: here?

As mentioned in another related comment, this is actually what our conventions as enforced by `rustfmt`/`rust-analyzer` dictate (though the configuration is disabled for some reason in this repo). There are other reasons why `self` is used, which I mentioned in my other comment, but in general it's good practice to use `self` like this rather than leaving them unqualified - both as a less ambiguous syntax, and to aid in skimming import lists to determine local vs external imports at a glance.

> Also, do we need to export ArtifactKind publicly? If so, maybe we should come up with a more "self-describing" name?

A couple of the lower-level public assembler APIs take this as an argument, so we need to export it. We could probably re-use `ModuleKind` though, however I wanted to distinguish the two because the artifacts produced by the assembler may be a different set than the types of modules that can be defined. I'm open to alternatives though. This kind of thing is typically called an "artifact" or "object" in compiler/assembler/linker contexts, and is inherently a bit general, but we might be able to be more specific in our case.